sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala

```diff
@@ -443,6 +453,22 @@ class HDFSFileCatalog(
     }
   }
 
+  /** List of files to consider when there is not inferred partitioning scheme */
+  private def unpartitionedDataFiles(): Seq[FileStatus] = {
+    // For each of the input paths, get the list of files inside them
+    paths.flatMap { path =>
+      // Make the path qualified (consistent with listLeafFiles and listLeafFilesInParallel).
+      val fs = path.getFileSystem(hadoopConf)
+      val qualifiedPath = path.makeQualified(fs.getUri, fs.getWorkingDirectory)
+
+      // If it is a directory (i.e. exists in leafDirToChildrenFiles), return its children files
+      // Or if it is a file (i.e. exists in leafFiles), return the path itself
+      leafDirToChildrenFiles.get(qualifiedPath).orElse {
```

Comment:
If we have multiple layers of sub-directories, seems `get(qualifiedPath)` will always be false?